actors/builtin/market/market_test.go

```diff
@@ -1 +1,347 @@
+	"context"
+	"testing"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/filecoin-project/specs-actors/actors/abi"
+	"github.com/filecoin-project/specs-actors/actors/abi/big"
+	"github.com/filecoin-project/specs-actors/actors/builtin"
+	"github.com/filecoin-project/specs-actors/actors/builtin/market"
+	"github.com/filecoin-project/specs-actors/actors/builtin/miner"
+	"github.com/filecoin-project/specs-actors/actors/runtime/exitcode"
+	"github.com/filecoin-project/specs-actors/actors/util/adt"
+	"github.com/filecoin-project/specs-actors/support/mock"
+	tutil "github.com/filecoin-project/specs-actors/support/testing"
+)
+
+func TestMarketActor(t *testing.T) {
+	marketActor := tutil.NewIDAddr(t, 100)
+	owner := tutil.NewIDAddr(t, 101)
+	provider := tutil.NewActorAddr(t, "provider")
+	providerIdAddr := tutil.NewIDAddr(t, 102)
+	worker := tutil.NewIDAddr(t, 103)
+	client := tutil.NewIDAddr(t, 104)
+	var st market.State
+
+	setup := func() (*mock.Runtime, *marketActorTestHarness) {
+		builder := mock.NewBuilder(context.Background(), marketActor).
+			WithCaller(builtin.SystemActorAddr, builtin.InitActorCodeID).
+			WithActorType(owner, builtin.AccountActorCodeID).
+			WithActorType(worker, builtin.AccountActorCodeID).
+			WithActorType(providerIdAddr, builtin.StorageMinerActorCodeID).
+			WithActorType(client, builtin.AccountActorCodeID)
+
+		rt := builder.Build(t)
+
+		rt.AddIDAddress(owner, owner)
+		rt.AddIDAddress(worker, worker)
+		rt.AddIDAddress(client, client)
+		rt.AddIDAddress(provider, providerIdAddr)
+
+		actor := marketActorTestHarness{t: t}
+		actor.constructAndVerify(rt)
+
+		return rt, &actor
+	}
+
+	t.Run("simple construction", func(t *testing.T) {
+		actor := market.Actor{}
+		receiver := tutil.NewIDAddr(t, 100)
+		builder := mock.NewBuilder(context.Background(), receiver).
+			WithCaller(builtin.SystemActorAddr, builtin.InitActorCodeID)
+
+		rt := builder.Build(t)
+
+		rt.ExpectValidateCallerAddr(builtin.SystemActorAddr)
+
+		params := adt.EmptyValue{}
+		ret := rt.Call(actor.Constructor, &params).(*adt.EmptyValue)
+
+		assert.Equal(t, adt.EmptyValue{}, *ret)
+		rt.Verify()
+
+		store := adt.AsStore(rt)
+
+		emptyMap, err := adt.MakeEmptyMap(store)
+		assert.NoError(t, err)
+
+		emptyArray, err := adt.MakeEmptyArray(store)
+		assert.NoError(t, err)
+
+		emptyMultiMap, err := market.MakeEmptySetMultimap(store)
+		assert.NoError(t, err)
+
+		var state market.State
+		rt.GetState(&state)
+
+		assert.Equal(t, emptyArray.Root(), state.Proposals)
+		assert.Equal(t, emptyArray.Root(), state.States)
+		assert.Equal(t, emptyMap.Root(), state.EscrowTable)
+		assert.Equal(t, emptyMap.Root(), state.LockedTable)
+		assert.Equal(t, abi.DealID(0), state.NextID)
+		assert.Equal(t, emptyMultiMap.Root(), state.DealIDsByParty)
+	})
+
+	t.Run("AddBalance", func(t *testing.T) {
+		t.Run("adds to provider escrow funds", func(t *testing.T) {
+			testCases := []struct {
+				delta int64
+				total int64
```

Comment:
Maybe for a follow-up, but a table-driven approach like this might work well for exercising all of the provider, owner, and worker as non-ID addresses as well as ID addresses.